internal/impl/kafka/enterprise/schema_registry_input.go

```diff
@@ -167,23 +174,57 @@ func (i *schemaRegistryInput) Connect(ctx context.Context) error {
 	i.connMut.Lock()
 	defer i.connMut.Unlock()
 
-	data, err := i.doSchemaRegistryRequest(ctx, i.url.JoinPath("subjects").String())
+	subjects, err := i.client.GetSubjects(ctx)
 	if err != nil {
 		return fmt.Errorf("failed to fetch subjects: %s", err)
 	}
 
-	var subjects []string
-	if err := json.Unmarshal(data, &subjects); err != nil {
-		return fmt.Errorf("failed to unmarshal HTTP response: %s", err)
-	}
-
 	i.subjects = make([]string, 0, len(subjects))
 	for _, s := range subjects {
 		if i.subjectFilter.MatchString(s) {
 			i.subjects = append(i.subjects, s)
 		}
 	}
 
+	if i.fetchInOrder {
+		schemas := map[int]schemaInfo{}
+		for _, subject := range i.subjects {
+			var versions []int
+			if versions, err = i.client.GetVersionsForSubject(ctx, subject); err != nil {
+				return fmt.Errorf("failed to fetch versions for subject %q: %s", subject, err)
+			}
+			if len(versions) == 0 {
+				i.mgr.Logger().Infof("Subject %q does not contain any versions", subject)
+				continue
+			}
+
+			for _, version := range versions {
+				var schema sr.SchemaInfo
+				if schema, err = i.client.GetSchemaBySubjectAndVersion(ctx, subject, &version); err != nil {
+					return fmt.Errorf("failed to fetch schema version %d for subject %q: %s", version, subject, err)
+				}
+
+				schemas[schema.ID] = schemaInfo{
+					SchemaInfo: schema,
+					subject:    subject,
+					version:    version,
+				}
```

Comment:
Thanks for spotting that! Indeed, this is a corner case that I was trying to get my head around and it looks like you're right and the schemas will get duplicate IDs :( The trouble is that I can't get Redpanda to import schemas with a fixed ID but different subjects:

```
$ curl -X POST -H "Content-Type: application/vnd.schemaregistry.v1+json" --data '{"id": 1, "version": 1, "schema": "{\"name\": \"Foo\", \"type\": \"record\", \"fields\": [{\"name\": \"data\", \"type\": \"int\"}]}"}' http://localhost:8081/subjects/foo/versions
{"id":1}%                                                                                                                                                                       
$ curl -X POST -H "Content-Type: application/vnd.schemaregistry.v1+json" --data '{"id": 1, "version": 1, "schema": "{\"name\": \"Foo\", \"type\": \"record\", \"fields\": [{\"name\": \"data\", \"type\": \"int\"}]}"}' http://localhost:8081/subjects/bar/versions
{"error_code":42207,"message":"Overwrite new schema with id 1 is not permitted."}%                                                                                              
```

It does work, however, if I drop the `id` field or both the `id and `version` fields:
```shell
$ curl -X POST -H "Content-Type: application/vnd.schemaregistry.v1+json" --data '{"schema": "{\"name\": \"Foo\", \"type\": \"record\", \"fields\": [{\"name\": \"data\", \"type\": \"int\"}]}"}' http://localhost:8081/subjects/bar/versions
{"id":1}%
$ curl -X POST -H "Content-Type: application/vnd.schemaregistry.v1+json" --data '{"version": 1, "schema": "{\"name\": \"Foo\", \"type\": \"record\", \"fields\": [{\"name\": \"data\", \"type\": \"int\"}]}"}' http://localhost:8081/subjects/baz/versions
{"id":1}%
```
However, if I do this, then it feels kind of brittle because I'd have to make sure I first send the schema which does have the `id` field set...

Reply:
This should be fixed now. I'll defer adding a test until Redpanda lifts the above constraint as part of 
CORE-7734.